src/trunkls.jl

```diff
@@ -48,55 +50,120 @@ x0 = [-1.2; 1.0]
 nls = ADNLSModel(F, x0, 2)
 stats = trunk(nls)
 # output
+"Execution stats: first-order stationary"
+```
+
+```jldoctest; output = false
+using JSOSolvers, ADNLPModels
+F(x) = [x[1] - 1.0; 10 * (x[2] - x[1]^2)]
+x0 = [-1.2; 1.0]
+nls = ADNLSModel(F, x0, 2)
+solver = TrunkSolverNLS(nls)
+stats = solve!(solver, nls)
+
+# output
+
 "Execution stats: first-order stationary"
 ```
 """
-function trunk(
+mutable struct TrunkSolverNLS{
+  T,
+  V <: AbstractVector{T},
+} <: AbstractOptimizationSolver
+  x::V
+  xt::V
+  temp::V
+  gx::V
+  gt::V
+  tr::TrustRegion{T, V}
+  Fx::V
+  Av::V
+  Atv::V
+end
+
+function TrunkSolverNLS(
+  nlp::AbstractNLPModel{T, V}
+) where {T, V <: AbstractVector{T}}
+  nvar = nlp.meta.nvar
+  nequ = nlp.nls_meta.nequ
+  x = V(undef, nvar)
+  xt = V(undef, nvar)
+  temp = V(undef, nvar)
+  gx = V(undef, nvar)
+  gt = V(undef, nvar)
+  tr = TrustRegion(gt, one(T))
+
+  Fx = V(undef, nequ)
+  Av = V(undef, nequ)
+  Atv = V(undef, nvar)
+  return TrunkSolverNLS{T, V}(x, xt, temp, gx, gt, tr, Fx, Av, Atv)
+end
+
+function LinearOperators.reset!(::TrunkSolverNLS) end
+
+@doc (@doc TrunkSolverNLS) function trunk(
   ::Val{:GaussNewton},
   nlp::AbstractNLSModel;
-  x::AbstractVector = copy(nlp.meta.x0),
+  x::V = nlp.meta.x0,
+  subsolver_type::Type{<:KrylovSolver} = CgSolver,
+  kwargs...,
+) where {V}
+  solver = TrunkSolverNLS(nlp)
+  return solve!(solver, nlp; x = x, kwargs...)
+end
+
+function SolverCore.solve!(
+  solver::TrunkSolverNLS{T, V},
+  nlp::AbstractNLPModel{T, V},
+  stats::GenericExecutionStats{T, V};
+  x::V = nlp.meta.x0,
   subsolver::Symbol = :lsmr,
-  atol::Real = √eps(eltype(x)),
-  rtol::Real = √eps(eltype(x)),
+  atol::Real = √eps(T),
+  rtol::Real = √eps(T),
   max_eval::Int = -1,
   max_time::Float64 = 30.0,
   bk_max::Int = 10,
   monotone::Bool = true,
   nm_itmax::Int = 25,
   trsolver_args::Dict{Symbol, Any} = Dict{Symbol, Any}(),
   verbose::Int = 0,
-)
+) where {T, V <: AbstractVector{T}}
   if !(nlp.meta.minimize)
     error("trunk only works for minimization problem")
   end
   if !unconstrained(nlp)
     error("trunk should only be called for unconstrained problems. Try tron instead")
   end
 
+  reset!(stats)
   start_time = time()
   elapsed_time = 0.0
 
+  solver.x .= x
+  x = solver.x
```

Comment:
```suggestion
```

Reply:
Sorry, just realize that this line is in fact necessary. Please ignore this comment.